src/main/java/net/glowstone/util/GlowTravelAgent.java

```diff
@@ -0,0 +1,324 @@
+        double minDistance = Double.MAX_VALUE;
+
+        final int blockX = destination.getBlockX();
+        final int blockZ = destination.getBlockZ();
+
+        for (int x = blockX - searchRadius; x < (blockX + searchRadius); x++) {
+            for (int z = blockZ - searchRadius; z < (blockZ + searchRadius); z++) {
+                for (int y = 127; y >= 0; y--) {
+                    final Block toCompare = world.getBlockAt(x, y, z);
+
+                    if (toCompare.getType() == Material.PORTAL
+                        && toCompare.getRelative(BlockFace.DOWN).getType() != Material.PORTAL
+                    ) {
+                        final Location location = toCompare.getLocation();
+                        double distance = RayUtil.getRayBetween(location, destination).length();
+                        if (distance < minDistance) {
+                            minDistance = distance;
+                            minLoc = location;
+                        }
+                    }
+                }
+            }
+        }
+        return minLoc;
+    }
+
+    @Override
+    public Location findOrCreate(Location location) {
+        Location found = findPortal(location);
+        if (found == null) {
+            if (canCreatePortal && createPortal(location)) {
+                found = findPortal(location);
+            } else {
+                found = location;
+            }
+        }
+        return found; // Bukkit JDoc says to return the input location if no portal could be found
+    }
+
+    @Override
+    public boolean getCanCreatePortal() {
+        return canCreatePortal;
+    }
+
+    @Override
+    public TravelAgent setCreationRadius(int creationRadius) {
+        // Prevent too-low values
+        this.creationRadius = creationRadius < 2 ? 0 : creationRadius;
+        return this;
+    }
+
+    @Override
+    public boolean createPortal(Location destination) {
+        if (world.getEnvironment() == World.Environment.THE_END) {
+            return false;
+        }
+        double distance = -1.0D;
+        //requested location data
+        final int reqX = destination.getBlockX();
+        final int reqY = destination.getBlockY();
+        final int reqZ = destination.getBlockZ();
+        //actual portal location
+        int actualX = reqX;
+        int actualY = reqY;
+        int actualZ = reqZ;
+        //axis of the portal. 0 -> X axis, 1, -1 -> Z axis
+        int axis = 0;
+        int random = ThreadLocalRandom.current().nextInt(4);
+        int worldHeight = world.getEnvironment() == World.Environment.NETHER ? 128 : 256;
+
+        //try to find a suitable area for building the portal in range between
+        // x - 16, z - 16 and x + 16, z + 16.
+
+        //first attempt: find a portal spot that not only guarantees space for the portal, but also
+        //space around it
+        // iteration from x - 16 to x + 16
+        for (int tempX = reqX - 16; tempX <= reqX + 16; ++tempX) {
```

Comment:
This method is huge. Do you think you can break it down to multiple methods that can be tested individually?

Reply:
@momothereal I don't really know. If you want me to I will wrap my head around this. There are only two blocks (the first two) that are similar enough to have me try to extract them to a method

Reply:
This probably needs to be rewritten with functional syntax; for example, the nested search loops for X and Z (which repeat each other a bit) could become a subroutine taking an IntBinaryOperator scoring function.

Reply:
the problem with this is that the whole thing is highly state dependent the way it is currently written and i will have to change that, which is hat i am currently working on

Reply:
It occurs to me that it might actually be better to combine the nested for-loops into one, which could be an easier place to optimize from and maybe even a candidate for ForkJoinPool parallelization.

Reply:
Never got into ForkJoinPool stuff. If it isn't that complicated I will learn it. If it isn't it might be an idea to merge this pr and then optimize it the way you want it.

Reply:
I don't think the ForkJoinPool stuff is in the scope of this PR. But improving the readability of the logic is a big plus.

Reply:
@mastercoms I think this was the last remaining comment to resolve

Reply:
Portal creation logic is inherently complex, I will let this slide for now.